x/lockup/keeper/lock_test.go

```diff
@@ -10,12 +10,119 @@ import (
 	"github.com/osmosis-labs/osmosis/osmomath"
 	cl "github.com/osmosis-labs/osmosis/v19/x/concentrated-liquidity"
 	cltypes "github.com/osmosis-labs/osmosis/v19/x/concentrated-liquidity/types"
+	"github.com/osmosis-labs/osmosis/v19/x/lockup/keeper"
 	"github.com/osmosis-labs/osmosis/v19/x/lockup/types"
 	lockuptypes "github.com/osmosis-labs/osmosis/v19/x/lockup/types"
 
 	sdk "github.com/cosmos/cosmos-sdk/types"
 )
 
+func (suite *KeeperTestSuite) TestRebondTokens() {
+	// coins that will be locked
+	lockedCoins := sdk.NewCoins(sdk.NewInt64Coin("stake", 10))
+
+	addr1 := suite.TestAccs[0]
+	defaultLockID := uint64(1)
+	nonExistingLockID := uint64(2)
+	notOwnerAddress := suite.TestAccs[1]
+
+	testCases := []struct {
+		name                string
+		unlock              bool
+		rebondLockID        uint64
+		createSyntheticLock bool
+		owner               sdk.AccAddress
+
+		expectedError error
+	}{
+		{
+			name:         "Valid test case",
+			unlock:       true,
+			owner:        addr1,
+			rebondLockID: defaultLockID,
+		},
+		{
+			name:          "Invalid: Trying to rebond a non existent lock id",
+			unlock:        true,
+			owner:         addr1,
+			rebondLockID:  nonExistingLockID,
+			expectedError: types.ErrLockupNotFound,
+		},
+		{
+			name:          "Invalid: Trying to rebond a non-unbonding lock",
+			owner:         addr1,
+			rebondLockID:  defaultLockID,
+			expectedError: types.ErrLockNotUnlocking,
+		},
+		{
+			name:                "Invalid: Trying to rebond a synthetic's underlying lock",
+			createSyntheticLock: true,
+			owner:               addr1,
+			rebondLockID:        defaultLockID,
+			expectedError:       types.ErrLockHasSyntheticLockup,
+		},
+		{
+			name:          "Invalid: Not the owner of a lock",
+			unlock:        true,
+			owner:         notOwnerAddress,
+			rebondLockID:  defaultLockID,
+			expectedError: types.ErrNotLockOwner,
+		},
+	}
+
+	for _, tc := range testCases {
+		suite.Run(tc.name, func() {
+			suite.SetupTest()
+
+			// lock coins
+			lockID := suite.LockTokens(addr1, lockedCoins, time.Second)
+
+			if tc.unlock {
+				// unlock coins
+				_, err := suite.App.LockupKeeper.BeginUnlock(suite.Ctx, lockID, lockedCoins)
+				suite.Require().NoError(err)
+			}
+
+			initialLock, err := suite.App.LockupKeeper.GetLockByID(suite.Ctx, lockID)
+			suite.Require().NoError(err)
+
+			// a sanity check to make sure the lock is in the correct state
+			if tc.unlock {
+				suite.Require().True(initialLock.IsUnlocking())
```

Comment:
Would you be able to add an additional set of test cases (or extend these ones by adding a new test field for time elapsed) that covers cases where time has elapsed since the lock started unlocking? Specifically, it is important we test behavior around when:
* the lock is partially finished unlocking (should function same as current tests)
* the lock is exactly finished unlocking (should not allow rebonding, as it should no longer be unlocking)
* the lock is past the unlocking period (should fail in all cases)

It would also be great to have cases covering behavior where `BeginUnlock` is run on part (not all) of `lockedCoins`, as this triggers a different logic branch that uses `SplitLock` that is currently untested.

Finally, we should also have test cases that include the creation of other locks and assert that they are unchanged.

Reply:
hey @AlpinYukseloglu! Those are great points, thank you for your review!

Note: I changed the logic of tests to make it as readable as possible. Now, in a test case struct I include so called `setupFunctions` that are being ran prior to testing. These include: [creation of initial lock](https://github.com/osmosis-labs/osmosis/pull/4713/files#diff-2b9fd6ced556a433e83723588bd6af7854355ceb79f1ff3d856de1eb75850797R42-R48), [unlocking the lock](https://github.com/osmosis-labs/osmosis/pull/4713/files#diff-2b9fd6ced556a433e83723588bd6af7854355ceb79f1ff3d856de1eb75850797R51-R54), [creating synthetic lockup](https://github.com/osmosis-labs/osmosis/pull/4713/files#diff-2b9fd6ced556a433e83723588bd6af7854355ceb79f1ff3d856de1eb75850797R51-R54), etc.

I addressed your comments in latest commits, here is a breakdown of everything for an easier review process:
1. [Partially finished unlocking lock](https://github.com/osmosis-labs/osmosis/pull/4713/files#diff-2b9fd6ced556a433e83723588bd6af7854355ceb79f1ff3d856de1eb75850797R68-R77)
2. [Exactly finished unlocking lock](https://github.com/osmosis-labs/osmosis/pull/4713/files#diff-2b9fd6ced556a433e83723588bd6af7854355ceb79f1ff3d856de1eb75850797R78-R91)
3. [Past unlocking period](https://github.com/osmosis-labs/osmosis/pull/4713/files#diff-2b9fd6ced556a433e83723588bd6af7854355ceb79f1ff3d856de1eb75850797R92-R105)
4. [BeginUnlock on some fraction of locked coins](https://github.com/osmosis-labs/osmosis/pull/4713/files#diff-2b9fd6ced556a433e83723588bd6af7854355ceb79f1ff3d856de1eb75850797R150-R155)
5. [Assertion that other locks are unchanged](https://github.com/osmosis-labs/osmosis/pull/4713/files#diff-2b9fd6ced556a433e83723588bd6af7854355ceb79f1ff3d856de1eb75850797R184-R187)

Though, I have one concern about exactly unlocked lock (2). (2) works because I run `EndBlocker` in the test prior to rebonding a lock. Because of that, it automatically removes the lock if the current block time is exactly the `endTime` of a lock (matured). 

However, in real scenario, when transactions of a block with block time equal to `endTime` of an unlocking lock are being executed, if someone tries to rebond this lock, they will be able to do so, since it was not yet removed from the store. I am a little confused as to why in case (2), we should fail? Since unlocking of maturing locks happens in `EndBlocker`, I do not see the reason for forbidding case (2) from passing

Reply:
tagging for re-review of the new test

cc: @mattverse @p0mvn

Reply:
Adding on case (2): basically, the test works right now by simulating an unreal scenario - by calling endBlocker and then running rebonding logic both with the same blocktime is essentially equivalent to a scenario, when two consecutive blocks have the same block time

Reply:
I think case 2 correct way would be being able to rebond if endblocker has not been called yet